### ISSUE TITLE
[v2] Navigating to previously visited pages with <Link> retains scroll position

### ISSUE THREAD

**Issue by robinpyon** [gatsbyjs/gatsby#7454]
## Description

Clicking on a `<Link>` in v2 (111) appears to scroll you to a previous scroll position (if it exists) rather than scroll you to the top.

### Steps to reproduce

I've been able to recreate this in `gatsby-starter-default`:
https://gatsby-v2-beta-111-scroll.netlify.com/

(The only change is artificial padding added to the first page above the `/page-2` `<Link>`)

- Visit https://gatsby-v2-beta-111-scroll.netlify.com/
- On the home page, scroll down to the `page-2` link and click it 
- On `page-2`, now click 'Go back to the homepage'

### Expected result

You are taken back to the top of the home page

### Actual result

You are taken back to the home page but scrolled down the page (at the exact same position you were when you linked the `page-2` link)

### Environment

```
System:
    OS: macOS High Sierra 10.13.6
    CPU: x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 8.11.1 - ~/.nvm/versions/node/v8.11.1/bin/node
    Yarn: 1.9.4 - ~/.nvm/versions/node/v8.11.1/bin/yarn
    npm: 5.6.0 - ~/.nvm/versions/node/v8.11.1/bin/npm
  Browsers:
    Chrome: 68.0.3440.106
    Firefox: 61.0.1
    Safari: 11.1.2
  npmPackages:
    gatsby: next => 2.0.0-beta.111
    gatsby-plugin-manifest: next => 2.0.2-beta.6
    gatsby-plugin-offline: next => 2.0.0-beta.9
    gatsby-plugin-react-helmet: next => 3.0.0-beta.4
  npmGlobalPackages:
    gatsby-cli: 1.1.58
```

### Other notes

- This behaviour doesn't occur on the v1 starter. 
- I understand that there was a switch to `reach-router`, could this be linked?
- Manually adding `window.scrollTo(0, 0) / window.scroll({top: 0})` in page `componentDidMount` had no effect, though it would work after wrapping it in an arbitrary `setTimeout` - though this yields an initial flash of content in the wrong scroll position.

By default, I imagine the desired behaviour would be for all clicked `<Link>` elements to scroll you to the top of that target page and that only using the browser back / forward buttons would retain scroll position state in this manner.

**Comment by Manoz** [comment]
Could be related to #7450 ?
I noticed the same behavior. I though it was a feature since it’s not totally stupid.
If I read a very long blog post and click on a link it’s nice to go back where I was without scrolling the whole page, even if it’s a weird behavior.

**Comment by colbyhub** [comment]
I'd rather it default to the top of the page rather than the previous scroll position, it would create a more predictable experience.
Could the `<Link>` component default to not scrolling to the previous scroll position, but still have it as an option via a `prop`?

**Comment by DylanVann** [comment]
Could be related: https://github.com/gatsbyjs/gatsby/issues/5656

The logic discussed for determining if something is a pop could be the issue. We want to restore scroll on pop (back button pushed), not if you're just going to a link you've been to before.

**Comment by Chuloo** [comment]
Verified, Here's a [simple repro using the Blog-Starter ](https://github.com/Chuloo/gatsby-issue-7454).
Clone repo and run `npm install` followed by `gatsby develop`.

**Comment by thebigredgeek** [comment]
So it seems like this *might* be blocked by a necessary API change to reach-router?  Otherwise it sounds like we have to keep a hash of seen routing keys in order to determine push vs pop

**Comment by humphreybc** [comment]
FYI I'm seeing this not just with previously visited links, but with all links.

**Comment by ryanwiemer** [comment]
I too can confirm this behavior when using gatsby`v2`. It is a particularly bad user experience when you have next/previous links at the bottom of the page as it keeps the user at bottom.

**Comment by jkimbo** [comment]
We came across this issue on https://www.zego.com and developed a work around by using our own custom `Link` component that sets a flag on the `window`:

```js
// components/Link.js
import { Link as BaseLink } from 'gatsby';

if (typeof window !== 'undefined') {
  window.__navigatingToLink = false;
}

const Link = ({ children, ...props }) => (
	<BaseLink onClick={() => { window.__navigatingToLink = true; }} {...props}>
		{children}
	</BaseLink>
);
```

Then in `gatsby-browser.js`:

```
// gatsby-browser.js
exports.shouldUpdateScroll = () => {
  if (window.__navigatingToLink === true) {
    return [0, 0];
  }
  return true;
};

exports.onRouteUpdate = () => {
  window.__navigatingToLink = false;
};
```

It's pretty hacky but it works. All our page links scroll to the top of the page like expected and the back button still maintains the scroll position on the previous page.

**Comment by CanRau** [comment]
great thanks for the tip @jkimbo that works 👍

**Comment by KyleAMathews** [comment]
@jkimbo great solution! I'll add that to core while we're awaiting a fix upstream in @reach/router.

**Comment by KyleAMathews** [comment]
Here's the link to the @reach/router issue https://github.com/reach/router/issues/119

**Comment by steverandy** [comment]
I'm using v2.0.0-rc.24 and it seems to be having similar issue again.

**Comment by ryanwiemer** [comment]
@KyleAMathews,

Did your temporary fix get reverted or canceled out somehow? I seeing the undesirable scrolling behavior again using `v2.0.0-rc.25`.

**Comment by ryanwiemer** [comment]
@davidbailey00 and @pieh,

Thanks for the update and background info! Hopefully this gets resolved in the near future as `v2` has been officially released. 

Let me know if you need help with anything.

**Comment by pieh** [comment]
Yeah, I probably broke it and will take a look on this

**Comment by cardiv** [comment]
I saw this behavior on [gatsbyjs.com](https://www.gatsbyjs.com/) today and found it quite nice. Could this be recreated for a case (and others) @Manoz mentioned? Or do you think it is totally user unfriendly?

**Comment by pieh** [comment]
@cardiv can you try adding this to your `gatsby-browser.js`:
```
exports.shouldUpdateScroll = ({
  routerProps: { location },
  getSavedScrollPosition
}) => {
  return getSavedScrollPosition(location)
}
```

**Comment by cardiv** [comment]
Thanks, @pieh! I will try that on occasion. 👍

**Comment by mrintoul** [comment]
Hi there, I believe I'm still having this issue in Gatsby 2.4.5. I've tried @jkimbo's solution but I'm still getting the same result. It can be seen on [this page of my site](https://deploy-preview-13--mrintoul.netlify.com/theswitch), at the bottom of the page when clicking to ROAR. When clicking back to The Switch, the scroll position is retained.

**Comment by CanRau** [comment]
I'm on Gatsby 2.0.62 now and my custom scroll behavior stopped working

Here it's still working, especially when navigating within the page it's scrolling to the main-nav to skip the header https://5c02ee4941174f0c53519200--gaiama.netlify.com/en it seems the upgrade from 2.0.31 to 2.0.60/62 broke it

Trying to figure out what changed, any hints?

**Comment by pieh** [comment]
@CanRau Do you use `gatsby-remark-autolink-headers` - if so this is very likely https://github.com/gatsbyjs/gatsby/pull/9657

**Comment by CanRau** [comment]
Yes, that sounds very likely, maybe it could be interesting for the plugin to offer an option to disable `shouldUpdateScroll`, and maybe even an importable function one can use in it's own `shouldUpdateScroll` implementation so you can decide for yourself how to prioritize?!
And besides this complexity I like your proposal 👍

**Comment by vinayan3** [comment]
Any updates on this? I've tried the hacks and it isn' working for me on the `"gatsby": "^2.0.75",'

**Comment by ar-stackrox** [comment]
Same here. On the latest version

**Comment by vinayan3** [comment]
@ar-stackrox I am digging through this again because the scrolling thing has been pretty irritating. I found out that `window.scrollTo` doesn't work when you have these following CSS statements
```
html, body { height: 100%; overflow:auto; }
```
My code base had that and I've changed it to

```
html {
  width: 100%;
  height: 100%;
  overflow: auto;
}
body {}
```

I found this solution on . https://stackoverflow.com/questions/1174863/javascript-scrollto-method-does-nothing

For now this seems to be a workaround where new pages now scrolled to the top.

**Comment by brookslyrette** [comment]
Seeing this as well on: `2.4.7`

**Comment by davidbailey00** [comment]
Hi everyone, I've just tested this out locally and can't see any problems with the current scroll behaviour. I can see the issue in the site @mrintoul, linked but this isn't occurring for me, so I'm suspecting it was fixed with a recent update.

From what I can tell there are some websites linked which highlight the problem, but reproduction repos linked in the comments - please could someone create a reproduction for us to investigate? I'd love to get this fixed if it's still an issue, so that using the `scrollTo` hack is unnecessary :)

**Comment by AllanPooley** [comment]
Hi guys, debugging this for a few hours last night, to no avail.

However, this morning I was blessed with an epiphany sent by the software engineering gods! 🕊️

I had some pretty funky styles to make a togglable navigation drawer / menu appear and disappear by pushing away the app's main containing element.

To make this work, I had added:

```
html {
  overflow: hidden;
}
```

In my case, removing this style fixed the behaviour!!

Hope this helps someone else.

**Comment by AllanPooley** [comment]
![image](https://user-images.githubusercontent.com/9204586/52456843-f36dba00-2baa-11e9-95e2-c07db6fcb2a1.png)

**Comment by thundernixon** [comment]
@davidbailey00 I'm currently having two pretty serious issues with scroll behavior, and this is one of them.

Here's a repo that reproduces this issue on gatsby `v2.0.118` (currently the latest version):

https://github.com/thundernixon/gatsby-link-scroll-test

Not only is it linking to previously-scrolled-to spots in pages (bad), but it’s also linking to the current location on *unvisited* pages (worse). That is, the pagination at the bottom of a “blog post” type of page will link to the bottom of the next page.

My second issue is described at https://github.com/gatsbyjs/gatsby/issues/11687.

**Comment by davidbailey00** [comment]
@thundernixon Thanks for your update! I'll presume there's no further issue

@AllanPooley Please could you share your repo, or a reproduction? It sounds like this could still be an issue, even if it only happens with the body styled as `overflow: hidden`, so ideally we'd like to get this fixed for all cases :)

**Comment by ergunpp** [comment]
Hi,
I have tried as @jkimbo with my project but I'm still having the same problem. I have added a Link.js under components, which I import to pages linking to other pages. I also added the code to my gatsby-browser.js. Same problem

**Comment by davidbailey00** [comment]
@ergunpp Please could you share your repo, or a minified reproduction, so we can investigate this?

**Comment by ergunpp** [comment]
@davidbailey00 repo name: ergunpp/chefschoice

**Comment by t2ca** [comment]
The same issue currently exists when using semantic-ui-react and it's because of the following css code:
```
body {
  height: 100%;
}
```
So i fixed it with the following:
```
html {
  height: initial;
}
```

**Comment by ergunpp** [comment]
Hi, my repo is ergunpp/chefschoice

**Comment by eagleeyejack** [comment]
Had this problem all day today digging through various threads looking for the answer, figured out a workaround even though it's not ideal.

I converted my index.js file into a class-based component and then on componentDidUpdate I added

`
  componentDidUpdate() {
    window.scrollTo(0,0);
  }
`

It's not pretty but now behaves the expected way. I have a feeling it's something to with my styles (overflow, height:100%, are some suggestions I've had from forums).

I could replicate the issue when I started adding my styles back in. 

Will post an update if I figure out the culprit.

**Comment by davidbailey00** [comment]
@ergunpp I just had a look at your site - it seems the problem is caused by your `overflow-x` and styling with the `body` tag. If you remove this property, as well as the `overflow-y` property, and then fix the page overflowing by changing various instances of `100vw` to `100%`, the problem is fixed :)

**Comment by davidbailey00** [comment]
I think after @pieh's fixes, any problems occurring are due to the `scroll-behavior` package, which is outside of our control, since the scroll update events seem to be firing correctly.

**Comment by KyleAMathews** [comment]
If scroll-behavior has problems, we can PR fixes there. We've done that several times.

**Comment by davidbailey00** [comment]
@KyleAMathews I think it's not even a problem with `scroll-behavior` - even calling `window.scrollTo(0, 0)` in the console doesn't work with some of those styles applied, so I think this isn't something we can fix, apart from telling people not to set `overflow: hidden`. Should we close?

**Comment by ergunpp** [comment]
> @ergunpp I just had a look at your site - it seems the problem is caused by your `overflow-x: hidden` and styling with the `body` tag. If you remove this property, as well as the `overflow-y` property, and then fix the page overflowing by changing various instances of `100vw` to `100%`, the problem is fixed :)

Hi @davidbailey00, Your suggestions worked like a charm. Thanks for the great help!

**Comment by davidbailey00** [comment]
Closing, since this has been fixed (or else caused by dodgy plugins / CSS). Please open a new issue if any further problems occur :)

**Comment by nandorojo** [comment]
This worked for me.

**To scroll to the top,** make sure your path starts with `/`. Example:

```javascript
<Link to="/some-path" />
```

I realized it wasn't scrolling to the top when I didn't have a `/` at the start like this: `<Link to="some-path" />

**Comment by mavam** [comment]
@t2ca your `html` CSS tweak fixed it for me after trying pretty much everything else here. 💯 

I'm also using Semantic UI React and have added the following to `site.overrides`:

```css
html {
  height: initial;
}
```

**Comment by rishichawda** [comment]
After wasting a whole hour going through all the issues and references, nothing worked except manually setting the `scrollTop` on `document.body` to 0 like so:

```js
export const shouldUpdateScroll = () => {
    document.body.scrollTop = 0
}
```

Works for both refresh and page visits.

**Comment by Uhiyamind** [comment]
This worked for me.

`exports.onInitialClientRender = () => {
  window.scrollTo(0, 0)
}
`

[https://github.com/gatsbyjs/gatsby/issues/19488#issuecomment-554007194](url)

**Comment by lvtz** [comment]
> This worked for me.
> 
> gatsby-browser.js
> 
> ```js
> exports.onInitialClientRender = () => {
>   window.scrollTo(0, 0)
> }
> ```

It does not work, and seems it can't - definition of this function looks like this: 

> Called when the initial (but not subsequent) render of Gatsby App is done on the client.

In my case it work great, no more scroll retain issue when navigating using Link from Gatsby.

`
// gatsby-browser.js

export const onPreRouteUpdate = () => {
  document.body.scrollTop = 0
}
`

**Comment by GesJeremie** [comment]
This worked pretty well:


```javascript
// gatsby-browser.js
export const shouldUpdateScroll = () => {
  return false
}
```

Source: https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#shouldUpdateScroll
